### PR TITLE
CAPX-43 Profile delete profile actions

### PR DESCRIPTION
**Pull request by sherakama**
# Ready For Review
- Created profile operations on delete importer confirmation
- Added confirmation messaging

### CODE REVIEW

**Comment by sherakama** [comment]
@mpriest @pingers 

Please have a review when you have a moment.

Steps to test:
1. Create mapper and importer and importer some profiles
2. Navigate to importer page admin/config/capx/importer
3. Click on delete action
4. Select operation from select menu on confirmation page
5. Delete the importer
6. Confirm action was applied
7. Rinse and repeat 1 - 6 for each action.

File: includes/CAPx/Drupal/Util/CAPx.php [hunk 1]
```diff
@@ -44,6 +44,64 @@ public static function getProfiles($type, $conditions = array()) {
...
+      throw new \Exception("Type required for getProfiles", 1);
+    }
+
+    $query = db_select("capx_profiles", 'capx')
+      ->fields('capx', array('entity_type', 'entity_id'))
+      ->condition('entity_type', $type)
+      ->orderBy('id', 'DESC');
+
+    foreach ($conditions as $key => $value) {
+      $query->condition($key, $value);
```
Comment (mpriest) [inline includes/CAPx/Drupal/Util/CAPx.php:97]: Ideally, should support operators for the conditions.

File: includes/CAPx/Drupal/Util/CAPx.php [hunk 2]
```diff
@@ -44,6 +44,64 @@ public static function getProfiles($type, $conditions = array()) {
...
+   *   An array of entity id integers.
+   */
+  public static function getProfileIds($type, $conditions = array()) {
+
+    if (!$type) {
+      throw new \Exception("Type required for getProfiles", 1);
+    }
+
+    $query = db_select("capx_profiles", 'capx')
+      ->fields('capx', array('entity_type', 'entity_id'))
```
Comment (mpriest) [inline includes/CAPx/Drupal/Util/CAPx.php]: Don't need to select "entity_type".

File: includes/CAPx/Drupal/Util/CAPx.php [hunk 3]
```diff
@@ -44,6 +44,64 @@ public static function getProfiles($type, $conditions = array()) {
...
+      ->condition('entity_type', $type)
+      ->orderBy('id', 'DESC');
+
+    foreach ($conditions as $key => $value) {
+      $query->condition($key, $value);
+    }
+
+    $result = $query->execute();
+    $assoc = $result->fetchAllAssoc('entity_id');
```
Comment (mpriest) [inline includes/CAPx/Drupal/Util/CAPx.php]: Can just use `$result->fetchCol();` to return a associative array and not worry about calling array_keys().

File: stanford_capx.forms.inc [hunk 4]
```diff
@@ -1169,41 +1169,98 @@ function stanford_capx_importer_form_submit_run_import($form, $form_state) {
...
     '#value' => $importer->machine_name,
   );
 
+  // Operations.
+  $form['profile_action'] = array(
+    '#type' => 'select',
+    '#title' => t('What would you like to do with the profiles that are associated with this importer?'),
+    '#description' => t('Perform an action on all of the profiles that have been imported by this importer.'),
+    '#options' => array(
+      'nothing' => 'Do nothing and leave the profiles as they are',
```
Comment (mpriest) [inline stanford_capx.forms.inc]: These should be translated.

File: stanford_capx.forms.inc [hunk 5]
```diff
@@ -1169,41 +1169,98 @@ function stanford_capx_importer_form_submit_run_import($form, $form_state) {
...
+  );
+
+  // Confirm form stuff.
   $question = "Are you sure you want to delete " . $importer->title . "?";
   $path = "admin/config/capx/importer";
   $description = t('This action cannot be undone.');
   $name = 'confirm_importer_delete';
-  $form = confirm_form($form, $question, $path, $description, 'Yes, please delete', 'Cancel', $name);
+
+  $form = confirm_form($form, $question, $path, $description, 'Yes, please delete the importer', 'Cancel', $name);
```
Comment (RoSk0) [inline stanford_capx.forms.inc]: Should use t() for strings that get shown to the user.

File: stanford_capx.forms.inc [hunk 6]
```diff
@@ -1169,41 +1169,98 @@ function stanford_capx_importer_form_submit_run_import($form, $form_state) {
...
   $importer = CAPxImporter::loadImporter($machine_name);
 
+  // No importer, no need to continue.
   if (!$importer) {
     throw new Exception("Could not load an importer with provided machine name. Please check url.");
   }
 
+  // Add the machine name so we can do stuff with the importer later.
   $form['machine-name'] = array(
     '#type' => 'hidden',
```
Comment (mpriest) [inline stanford_capx.forms.inc]: This should be `'#type' => 'value',`.

File: stanford_capx.forms.inc [hunk 7]
```diff
@@ -1169,41 +1169,98 @@ function stanford_capx_importer_form_submit_run_import($form, $form_state) {
...
+ * profiles.
+ *
  * @param  [type] $form       [description]
  * @param  [type] $form_state [description]
  * @return [type]             [description]
  */
 function stanford_capx_admin_config_importer_delete_submit($form, $form_state) {
-  $importer = CAPxImporter::loadImporter($form_state['values']['machine-name']);
+
+  $machine_name = check_plain($form_state['values']['machine-name']);
```
Comment (mpriest) [inline stanford_capx.forms.inc]: If we change the form element type to value, we don't need to check_plain().

File: stanford_capx.forms.inc [hunk 8]
```diff
@@ -1169,41 +1169,98 @@ function stanford_capx_importer_form_submit_run_import($form, $form_state) {
...
+  $type = $entity_importer->getEntityType();
+  $importer = $entity_importer->getImporter();
+  $profile_op = check_plain($form_state['values']['profile_action']);
+
+  switch ($profile_op) {
+    case 'unpublish':
+      $profiles = CAPx::getProfilesByImporter($machine_name);
+      $count = count($profiles);
+      foreach ($profiles as $profile) {
+        $profile->status = 0;
```
Comment (mpriest) [inline stanford_capx.forms.inc:1251]: This is node specific - we can import to things that are not nodes (taxonomy and users).

File: stanford_capx.forms.inc [hunk 9]
```diff
@@ -1169,41 +1169,98 @@ function stanford_capx_importer_form_submit_run_import($form, $form_state) {
...
+        $profile->save();
+      }
+      drupal_set_message($count . ' profiles were set to unpublished.', 'status');
```
Comment (mpriest) [inline stanford_capx.forms.inc]: Same as above. Node specific.

File: stanford_capx.forms.inc [hunk 10]
```diff
@@ -1169,41 +1169,98 @@ function stanford_capx_importer_form_submit_run_import($form, $form_state) {
...
+
+      break;
+
+    case "delete":
+      $ids = CAPx::getProfileIds($type, array('importer' => $machine_name));
+      $count = count($ids);
+      entity_delete_multiple($type, $ids);
+      drupal_set_message($count . ' profiles were deleted.', 'status');
```
Comment (mpriest) [inline stanford_capx.forms.inc]: Should probably use the entity (or bundle) name in place of "profiles". You can get this from the entity info I.e. type or 'plural label'.

File: stanford_capx.forms.inc [hunk 11]
```diff
@@ -1169,41 +1169,98 @@ function stanford_capx_importer_form_submit_run_import($form, $form_state) {
...
+  $importer = $entity_importer->getImporter();
+  $profile_op = check_plain($form_state['values']['profile_action']);
+
+  switch ($profile_op) {
+    case 'unpublish':
+      $profiles = CAPx::getProfilesByImporter($machine_name);
+      $count = count($profiles);
+      foreach ($profiles as $profile) {
+        $profile->status = 0;
+        $profile->save();
```
Comment (mpriest) [inline stanford_capx.forms.inc]: Need to do the following:

``` php
$wrapper = entity_metadata_wrapper($type, $profile);
$wrapper->save();
```

We don't actually have a wrapper at this point, so we don't have save() method available.

**Comment by mpriest** [comment]
Field processor logic is flawed - it's something I'm working on in the validation. Need to handle required fields using default values - not setting null. Also need to better detect when the incoming data is not null.

Once I got past those hurdles on this PR, testing went well - except for unpublishing. This just needs a change to load the profile in the wrapper to get access to the save() method. That's the only critical fix required.

The other general comment is that "unpublishing" is a node specific term. Should probably implement some logic around the entity type supported by the mapper to display appropriate language and actions. Users and nodes have "status", but taxonomy terms do not. Unpublish action won't have any effect on terms.

**Comment by sherakama** [comment]
Wow, thanks for all the comments. I will look into them and provide feedback on each. 

The logic for unpublishing is broken. I think for first pass I will do a check for the entity type of node and provide that option. For everything else it is leave it or delete it.

File: includes/CAPx/Drupal/Util/CAPx.php (hunk 1 above)
Comment (sherakama) [inline includes/CAPx/Drupal/Util/CAPx.php:97]: This is meant to be a quick helper for actually writing the query, I have added another function parameter to support this as an all or nothing for now. If a greater use for this is needed we can refactor.

File: includes/CAPx/Drupal/Util/CAPx.php (hunk 2 above)
Comment (sherakama) [inline includes/CAPx/Drupal/Util/CAPx.php]: Good catch.

File: includes/CAPx/Drupal/Util/CAPx.php (hunk 3 above)
Comment (sherakama) [inline includes/CAPx/Drupal/Util/CAPx.php]: Thanks for this. Good savings.

File: stanford_capx.forms.inc (hunk 4 above)
Comment (sherakama) [inline stanford_capx.forms.inc]: Good catch.

File: stanford_capx.forms.inc (hunk 6 above)
Comment (sherakama) [inline stanford_capx.forms.inc]: I've never used this form api element before. Thanks for the tip.

File: stanford_capx.forms.inc (hunk 8 above)
Comment (sherakama) [inline stanford_capx.forms.inc:1251]: Both users and nodes have a status attribute. I've provided a check to ensure that only nodes and users have this option available to them.

File: stanford_capx.forms.inc (hunk 10 above)
Comment (sherakama) [inline stanford_capx.forms.inc]: I agree on the use of the word "Profiles"

I went generic with the entity type for now.

File: stanford_capx.forms.inc (hunk 11 above)
Comment (sherakama) [inline stanford_capx.forms.inc]: Thanks for this.

**Comment by sherakama** [comment]
There was no major design issues posted so I am going to rebase and merge in now that I have fixed all of your above comments. 

Thanks again for the help.

**Pull request merged**